### ISSUE TITLE
Create a view to list and review unused lemmas

### ISSUE THREAD

**Issue by Woseseltops** [Signbank/Global-signbank#733]
I reviewed the code but could not find this functionality already. I suggest a list with a 'delete' button behind each item.

Requested by Hannah Lutzenberger

**Comment by ocrasborn** [comment]
Would it be an idea to integrate this in #529, that is, to add a column to an import file that does not match a field in the database but that could be used to mark Lemma ID Glosses for deletion?
In the CSV export of Lemma ID Glosses, it would be helpful then to add a column that lists (comma-separated) the Annotation ID Glosses that *are* linked to a lemma; with that, users could easily search for the unlinked ones.

**Comment by susanodd** [comment]
If you are on the Lemma List view, you see a column with Number of glosses.
If there are no glosses, there is a trash bin in the Update column

**Comment by hannahlutzenberger** [comment]
susanodd, this is what I have been using, but it is quite time-consuming to delete unused Lemmas in this dataset (1636 lemmas on 33 pages). What I was interested in is whether there is a more efficient way to this, i.e. bunch deleting the lemmas with trash bin symbol (or even better, an automatically deleting unused lemmas).

**Comment by susanodd** [comment]
Okay, thanks for the feedback. That is way too many to delete by hand.
Originally, the lemma's were created by default because the field idgloss used to be in the Gloss model. So those got turned into  lemma's behind the scenes. The lemma's also are where the (pointer to the) Dataset resides.
I will look to see what is going on. It looks like things got made obsolete in some way.

**Comment by hannahlutzenberger** [comment]
yes, initially they were created automatically but I am using them to further group sign variants together in the Kata Kolok dataset. This means that I may create new ones and edit existing ones to collect a few variants. It may not be an issue in the NGT dataset  but for me it would be quite helpful! thank you!

**Comment by susanodd** [comment]
I added numerous improvements to Admin LemmaIdgloss and LemmaIdglossTranslation in #739. Those aren't online yet. There are a couple of screenshot examples of Kata Kolok. Although those are focussing on administration of the Lemma objects, not on whether they are used.

I'm working on this issue now.

**Comment by Woseseltops** [comment]
Curious what you are building @susanodd . To give you some inspiration, maybe
* an overview of all unused lemmas with checkboxes before them that are checked by default
* a button 'delete all selected'

**Comment by susanodd** [comment]
So far, I'm adding an additional "form your query" parameter, No Glosses.
Then all the results will have no glosses.
Okay, a column would work.

**Comment by susanodd** [comment]
Here's what it looks like so far. It's kind of terrifying. There are 723 lemma's without glosses. There are 15 pages of results.
Does a select column seem wise? In Dutch: het is niet overzichtelijk.


![lemma-query-only-no-glosses](https://user-images.githubusercontent.com/24272544/116696850-b5932100-a9c2-11eb-8533-3d00ec7e9250.png)

**Comment by susanodd** [comment]
What's the most user-friendly, least pannicky way to delete lots of objects?

**Comment by susanodd** [comment]
I'll see what I can come up with. It makes me a bit nervous if a user wants to delete so many objects at once.
I see above you suggested adding a column in the CSV export for marking. Although that would also be the same, except it wouldn't be executed right away.
Maybe there is a way to "schedule for delete" and not physically delete them for a while?
I guess if the user first wants to export everything (make a backup) and then delete them?
Perhaps that is what you are imagining doing.

So suppose the user plans to delete all of these. What additional information should be exported in a "schedule for deletion" CSV, before actually deleting?

**Comment by susanodd** [comment]
I added the opposite toggle, to only show lemmas that have glosses.

![lemmas-katakolok-has-glosses](https://user-images.githubusercontent.com/24272544/116875748-b377cf00-ac1b-11eb-8388-c88ed003f192.png)

**Comment by susanodd** [comment]
I can add the "num_gloss" column to the CSV export.

Above, @ocrasborn asks https://github.com/Signbank/Global-signbank/issues/733#issuecomment-806814815

One glitch, in which language should the gloss's Annotation appear in the CSV?

> In the CSV export of Lemma ID Glosses, it would be helpful then to add a column that lists (comma-separated) the Annotation ID Glosses that _are_ linked to a lemma; with that, users could easily search for the unlinked ones.

**Comment by susanodd** [comment]
@ocrasborn would you like an additional filter on whether translations are missing, or complete?
It's not currently possible to search on an empty translation.

**Comment by ocrasborn** [comment]
> @ocrasborn would you like an additional filter on whether translations are missing, or complete?
It's not currently possible to search on an empty translation.

Let's not for now, there are enough other high priority issues to deal with first. Missing translations can be easily detected in a CSV export.

**Comment by susanodd** [comment]
I added a Number of glosses column to the CSV output.

**Comment by susanodd** [comment]
Does this look okay?

![deletelemmaswithnoglossesbutton](https://user-images.githubusercontent.com/24272544/117280195-a4369280-ae62-11eb-8ecd-9a4e0d895101.png)

**Comment by susanodd** [comment]
I prefer to do it with a button because when there are numerous pages of lemma's with no glosses, toggling a select for deletion column would be excessive.
Incidentally, I can't find a non-Latin spelling of lemma (plural).

**Comment by ocrasborn** [comment]
Yes @susanodd , looks OK to me. I do notice that we're very unsystematic with the capitalisation of words, here even within a single button! Perhaps just revise the glosses into Glosses in the button on the right for now, but I'll create an issue for us to look at this more systematically at some point.

**Comment by susanodd** [comment]
@vanlummelhuizen @Woseseltops @ocrasborn Q:

For deleting lemmas, at the moment, the permission: 
        

>    `dictionary.delete_lemmaidgloss`

is checked, as well as that the user is logged in.
Is this sufficient?

**Comment by susanodd** [comment]
The tests all run fine.

**Comment by susanodd** [comment]
Here are some screenshots showing the deletion with confirm, on a restricted query.

![lemma-list-kata-kolok-chop](https://user-images.githubusercontent.com/24272544/117956917-ce33fd00-b319-11eb-94c7-89a31ed7f618.png)

![lemma-list-kata-kolok-confirm](https://user-images.githubusercontent.com/24272544/117956936-d3914780-b319-11eb-9ceb-0a6607815b07.png)

![lemma-list-search-chop-after-delete](https://user-images.githubusercontent.com/24272544/117956959-d9872880-b319-11eb-95f4-acd2af74e81f.png)

![lemma-list-kata-kolok-other-no-glosses](https://user-images.githubusercontent.com/24272544/117956983-dee47300-b319-11eb-986f-dfb56668d098.png)

**Comment by susanodd** [comment]
Revised the lemma search so it works with a regular expression, like "^CH".

**Comment by susanodd** [comment]
Okay, I've added the change_dataset permission check 

There are already permission checks on the query that fetches the lemma objects.
The deletes are done inside a post method.
I added the (extra) permission checks there.

I suspect that the only way I can test the permissions is by creating a test that can do this. It seems like an artificial case scenario.

As a user, I don't see how I can obtain a set of lemma objects outside of the interface and get the post method to operate on them anyway.
I'll implement a test!

[There is no direct URL for deleting the set of lemmas. It's in the post method. That's tied to the query. But it's true that there are some computations done on this query, namely checking the count of glosses. And the delete button checks `dictionary.delete_lemmaidgloss` as described above.)

**Comment by susanodd** [comment]
I have the first test done. It creates a lemma with a gloss and a lemma without a gloss. Then it systematically tries out not having the required permissions.
The next test will create more lemmas to insure that a lemma (without glosses) not matching a query is not deleted.

**Comment by susanodd** [comment]
Permission checks vamped up. Implemented necessary tests.
This has been pushed to branch `lemmas_minimalpairs_733_74`
Awaiting code review.

**Comment by vanlummelhuizen** [comment]
> As a user, I don't see how I can obtain a set of lemma objects outside of the interface and get the post method to operate on them anyway.

If a user can guess object IDs outside the list (s)he retrieves, (s)he could construct and send a POST request using guessed IDs. Since Signbank uses consecutive integers as IDs, those IDs are easily guessed.

So, good that you also implemented the change_dataset permission!

**Comment by susanodd** [comment]
Gee, I certainly hope we don't have users doing that! Yykes!

**Comment by susanodd** [comment]
I tried out adding a logged in check, but that didn't do much because Django won't go to the url anyway, if the user is not logged in. It's protected in the urls.py file. (The attempt to test didn't succeed because it can't go to the url.)

**Comment by Woseseltops** [comment]
Hey @susanodd , sorry for the late response from my side. I really like how you solved this. I think it's good there is an extra alert before you remove lemmas! 

Two things:
* I'd make the initial button red, to convey danger (and other destroy buttons in Signbank are red as well), and make the cancel button in the dialogue a neutral color (really interesting: https://ux.stackexchange.com/questions/49991/should-yes-delete-it-be-red-or-green).
* Perhaps a small extra sanity check for the end user, to prevent mistakes, could be to add somewhere how many lemmas will be deleted if you click the delete button? Could be in the first button, or in the dialogue, or... whatever is easiest to realize :) .

**Comment by susanodd** [comment]
Okay, I've made the initial button red.
On the dialog buttons, they are the same as the dialog for deleting an individual row (green Cancel red Confirm Delete).

Okay, I finished and added the number of lemma's to be deleted.
Then I accidentally hit Confirm so my test lemma's disappeared from tstMH.

I checked all combinations of input using Kata Kolok. I made it so if there are no lemmas without glosses, the initial Delete button isn't shown.

Here is what it looks like for a query of ^BU

![new-lemma-nogloss-delete-buttons](https://user-images.githubusercontent.com/24272544/119971074-ce3a2b00-bfb0-11eb-824f-4b5a0362f26f.png)
